spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/security/oauth2/client/OAuth2ClientAutoConfigurationTests.java

```diff
@@ -0,0 +1,630 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.autoconfigure.security.oauth2.client;
+
+import java.lang.reflect.Field;
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import javax.servlet.Filter;
+
+import org.junit.After;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.NoSuchBeanDefinitionException;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.context.PropertyPlaceholderAutoConfiguration;
+import org.springframework.boot.autoconfigure.security.SecurityAutoConfiguration;
+import org.springframework.boot.context.config.ConfigFileApplicationListener;
+import org.springframework.boot.context.event.ApplicationEnvironmentPreparedEvent;
+import org.springframework.boot.test.util.TestPropertyValues;
+import org.springframework.context.annotation.AnnotationConfigApplicationContext;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.authentication.ProviderManager;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.oauth2.client.authentication.AuthorizationCodeAuthenticationProcessingFilter;
+import org.springframework.security.oauth2.client.authentication.AuthorizationCodeAuthenticationProvider;
+import org.springframework.security.oauth2.client.authentication.AuthorizationCodeRequestRedirectFilter;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
+import org.springframework.security.oauth2.client.user.nimbus.NimbusOAuth2UserService;
+import org.springframework.security.oauth2.core.AuthorizationGrantType;
+import org.springframework.security.oauth2.core.ClientAuthenticationMethod;
+import org.springframework.security.web.DefaultSecurityFilterChain;
+import org.springframework.security.web.FilterChainProxy;
+import org.springframework.security.web.util.matcher.AnyRequestMatcher;
+import org.springframework.util.ReflectionUtils;
+import org.springframework.web.context.support.AnnotationConfigWebApplicationContext;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Tests for {@link OAuth2ClientAutoConfiguration}.
+ *
+ * @author Joe Grandja
+ */
+public class OAuth2ClientAutoConfigurationTests {
```

Comment:
These tests would benefit from using our new `ApplicationContextTester`. We could take care of that as part of merging things. It may solve the `@EnableWebSecurity`-related problem as, among other things, it takes care of ordering auto-configuration classes correctly.

Reply:
I changed the tests to ensure `AnnotationConfigWebApplicationContext.register()` is called in the correct order for the auto-config classes. All tests are passing now. Let me see if I can integrate `WebApplicationContextRunner` instead.

Reply:
I tried integrating `WebApplicationContextRunner`, however, I couldn't get it working because `ClientPropertyDefaultsEnvironmentPostProcessor` did not `postProcess`.

Looking at the API, it wasn't clear on how to _register_ `ClientPropertyDefaultsEnvironmentPostProcessor` with the runner.

Reply:
Sounds like another reason not to use the environment post processor approach to me. Another is that you can't switch it off, i.e. you'll always get the property source in the environment, whether or not you're using Spring Security.

I don't think we'll be comfortable merging this until the use of an EPP has been replaced with something else.

Reply:
The property source will only be added if there are client properties available in the `Environment` if configured by the user. Otherwise, it won't be added.

Reply:
Thanks for the correction. I'd missed that. Unfortunately, it doesn't change my opinion about using an EnvironmentPostProcessor here. It would be the only auto-configuration in Boot that uses one. We're consistent in using `@ConfigurationProperties` for this pretty much everywhere else in Boot and I'd like that to continue. Can you please explore the `Map<String, Foo>` approach that I described earlier?

Reply:
See [comment](https://github.com/spring-projects/spring-boot/pull/10235#discussion_r138113710)